### ISSUE TITLE
IllegalStateException when deleting consumer while polling for records

### ISSUE THREAD

**Issue by ppatierno** [strimzi/strimzi-kafka-bridge#255]
Having an HTTP consumer polling for records and suddenly sending a DELETE request for removing itself from the bridge, can drive to the following error.

```
2019-06-17 11:36:16,695 [     vert.x-eventloop-thread-3] HttpBridgeContext              INFO  OpenAPI operation = poll
2019-06-17 11:36:16,695 [     vert.x-eventloop-thread-3] HttpSinkBridgeEndpoint         DEBUG [260617538] Request: body = null
2019-06-17 11:36:16,754 [     vert.x-eventloop-thread-3] HttpUtils                      INFO  [982326916] Request: method = DELETE, path = /consumers/my-group/instances/my-consumer
2019-06-17 11:36:16,754 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [982326916] Request: headers = user-agent: Vert.x-WebClient/3.7.1
content-type: application/vnd.kafka.v2+json
host: new-host-3:8080

2019-06-17 11:36:16,754 [     vert.x-eventloop-thread-3] HttpBridgeContext              INFO  OpenAPI operation = deleteConsumer
2019-06-17 11:36:16,755 [     vert.x-eventloop-thread-3] HttpSinkBridgeEndpoint         DEBUG [982326916] Request: body = null
2019-06-17 11:36:16,756 [     vert.x-eventloop-thread-3] HttpSinkBridgeEndpoint         INFO  Deleted consumer my-consumer from group my-group
2019-06-17 11:36:16,756 [     vert.x-eventloop-thread-3] HttpUtils                      INFO  [982326916] Response: statusCode = 204, message = No Content
2019-06-17 11:36:16,756 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [982326916] Response: headers = 
2019-06-17 11:36:16,757 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [982326916] Response: body = []
2019-06-17 11:36:16,761 [vert.x-kafka-consumer-thread-0] AbstractCoordinator            INFO  [Consumer clientId=my-consumer, groupId=my-group] Sending LeaveGroup request to coordinator new-host-3.station:9092 (id: 2147483647 rack: null)
Jun 17, 2019 11:36:16 AM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.lang.IllegalStateException: Response has already been written
	at io.vertx.core.http.impl.HttpServerResponseImpl.checkValid(HttpServerResponseImpl.java:614)
HttpServerResponseImpl.java:614
	at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:193)
HttpServerResponseImpl.java:193
	at io.strimzi.kafka.bridge.http.HttpUtils.sendResponse(HttpUtils.java:23)
HttpUtils.java:23
	at io.strimzi.kafka.bridge.http.HttpSinkBridgeEndpoint.lambda$6(HttpSinkBridgeEndpoint.java:250)
HttpSinkBridgeEndpoint.java:250
	at io.strimzi.kafka.bridge.SinkBridgeEndpoint.handleManualRecords(SinkBridgeEndpoint.java:683)
SinkBridgeEndpoint.java:683
	at io.strimzi.kafka.bridge.SinkBridgeEndpoint.lambda$16(SinkBridgeEndpoint.java:705)
SinkBridgeEndpoint.java:705
	at io.vertx.kafka.client.consumer.impl.KafkaConsumerImpl.lambda$poll$16(KafkaConsumerImpl.java:560)
KafkaConsumerImpl.java:560
	at io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl.lambda$null$40(KafkaReadStreamImpl.java:731)
KafkaReadStreamImpl.java:731
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
ContextImpl.java:320
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
EventLoopContext.java:38
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
AbstractEventExecutor.java:163
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
SingleThreadEventExecutor.java:404
```

After consumer deletion, the channel is closed and the bridge is trying to reply to the poll request.

**Comment by ppatierno** [comment]
This issue drove me to investigate more and doing more tests with a demo app I am developing with consumer polling the bridge periodically. At some point, the client (when killed) shutdown gracefully sending a delete consumer request first and waiting for the answer.

It is always fine when the polling interval is higher the poll timeout (the time spent by the Kafka native consumer in the bridge on the poll call waiting for available records, then returning no records).
In this scenario, the bridge is able to reply to the client before getting another poll request from it.

Problems started to show up when the client uses a poll interval less than poll timeout (i.e. polling every 1000 ms but with a timeout of 2000 ms) and there are no records available (so it means that the native poll runs for all the timeout value before returning).

The PR #256 related to this issue mainly fixes two things:

* the `routingContext` in the sink/source endpoint instances is captured by "old" poll callback but then they use a new routing context instance because it's overridden by new incoming polls.
* an "old" poll was going to use a "new" poll callback when the native poll unlock due to an instance member used as handler/callback which is overridden every new incoming poll as well.

I did a couple of tests with following results.

Without having HTTP client "pipelining":
* in this case, every new request is done on a different socket. The client sends every poll interval ms but anyway it sends a new poll on a different socket even if it hasn't received yet a response to the previous one. It's "pipelining" not in the HTTP 1.1 sense but just sending polls on different sockets without waiting for replies.
* in this scenario it can happen that when the delete request comes to the bridge and it replies, the client ends. In this case, the client could not get one (or more) response for previous polls (just before the delete) because they were still blocked on the native poll (due to the higher timeout). This is of course expected.

Having the HTTP client "pipelining":
* in this case, every new request is done on the same socket (even if it also depends on the max number of pipelined requests which is configurable, which depends then on how often poll is done and how is the poll timeout). Btw the client sends poll requests without waiting for responses.
* in this scenario, the bridge will handle the pipelining (thanks to the underlying router component) and will always reply in order. It means that even when the delete consumer request comes, the bridge handle or the callback for polls first (back from the native poll) and finally run the delete and replies to the client. It means that the client gets all responses for the polls and then the one for the delete.

**Comment by tombentley** [comment]
>Without having HTTP client "pipelining":
>
>* in this case, every new request is done on a different socket. The client sends every poll interval ms but anyway it sends a new poll on a different socket even if it hasn't received yet a response to the previous one. It's "pipelining" not in the HTTP 1.1 sense but just sending polls on different sockets without waiting for replies.
>* in this scenario it can happen that when the delete request comes to the bridge and it replies, the client ends. In this case, the client could not get one (or more) response for previous polls (just before the delete) because they were still blocked on the native poll (due to the higher timeout). This is of course expected.

I don't understand what you mean by "the client ends", but if each request from the client is being done from a different socket then the bridge should really be returning a response to a each poll (i.e. on each socket). Obviously it would have to be an error response to say that the consumer had been closed. If the bridge can arrange to close the TCP connection for the poll then it should be able to send a response too. Think about it from the pov of someone writing a client: It will be difficult to debug their client if some threads just observe the TCP connection being closed without an HTTP response. Is there some technical reason why we can't send a HTTP response on the HTTP connections for the pol requests?

**Comment by ppatierno** [comment]
... because by "the client ends" I mean that in the application I am using for testing, on the shutdown hook, the application waits until the consumer is deleted properly but it exits after a timeout (so the client connections are just closed).
See here:

https://github.com/ppatierno/kafka-http-strimzi-apps/blob/master/http-consumer/src/main/java/io/ppatierno/kafka/App.java#L55

and

https://github.com/ppatierno/kafka-http-strimzi-apps/blob/master/http-consumer/src/main/java/io/ppatierno/kafka/HttpKafkaConsumer.java#L68

If I kill the application, the shutdown hook undeploys the consumer verticle which stops the timer about sending poll requests and calls the delete consumer. At this point in the shutdown, I wait that the delete is done properly with a timeout: doing more long poll requests in the above scenarios it can happen that the delete is postponed so the application just ends for timeout on the latch.
